src/core/MIGRATION.md

```diff
@@ -0,0 +1,683 @@
+  }
+
+  public setup(core: CoreSetup) {
+    // called when plugin is setting up during Kibana's startup sequence
+  }
+
+  public stop(core: PluginStop) {
+    // called when plugin is torn down during Kibana's shutdown sequence
+  }
+}
+```
+
+The platform does not impose any technical restrictions on how the internals of the plugin are architected, though there are certain considerations related to how plugins interact with core and how plugins interact with other plugins that may greatly impact how they are built.
+
+### Services
+
+The various independent domains that make up `core` are represented by a series of services, and many of those services expose public interfaces that are provided to _all_ plugins via the first argument of their `setup` and `stop` functions. The interface varies from service to service, but it is always accessed through this argument.
+
+For example, the core `UiSettings` service exposes a function `get` to all plugin `setup` functions. To use this function to retrieve a specific UI setting, a plugin just accesses it off of the first argument:
+
+```ts
+import { CoreSetup } from '../../../core/public';
+
+export class Plugin {
+  public setup(core: CoreSetup) {
+    core.uiSettings.get('courier:maxShardsBeforeCryTime');
+  }
+}
+```
+
+Different service interfaces can and will be passed to `setup` and `stop` because certain functionality makes sense in the context of a running plugin while other types of functionality may have restrictions or may only make sense in the context of a plugin that is stopping.
+
+For example, the `stop` function in the browser gets invoked as part of the `window.onbeforeunload` event, which means you can't necessarily execute asynchronous code here in a reliable way. For that reason, `core` likely wouldn't provide any asynchronous functions to plugin `stop` functions in the browser.
+
+### Integrating with other plugins
+
+Plugins can expose public interfaces for other plugins to consume. Like `core`, those interfaces are bound to `setup` and/or `stop`.
+
+Anything returned from `setup` or `stop` will act as the interface, and while not a technical requirement, all Elastic plugins should expose types for that interface as well.
+
+**foobar plugin.ts:**
+
+```ts
+export type FoobarPluginSetup = ReturnType<Plugin['setup']>;
+export type FoobarPluginStop = ReturnType<Plugin['stop']>;
+
+export class Plugin {
+  public setup() {
+    return {
+      getFoo() {
+        return 'foo';
+      }
+    };
+  }
+
+  public stop() {
+    return {
+      getBar() {
+        return 'bar';
+      }
+    }
+  }
+}
+```
+
+Unlike core, capabilities exposed by plugins are _not_ automatically injected into all plugins. Instead, if a plugin wishes to use the public interface provided by another plugin, they must first declare that plugin as a dependency in their `kibana.json`.
+
+**demo kibana.json:**
+
+```json
+{
+  "id": "demo",
+  "requiredPlugins": [
```

Comment:
Renaming this configuration key to `dependencies` or `pluginDependencies` will make it more obvious that this is the source of the `dependencies` argument in `public setup(core: CoreSetup, dependencies: DemoSetupDependencies) {...`

Reply:
I'm going to leave this as-is since this is how it exists today, but I'm going to update this guide to use the word `plugins` instead of `dependencies` to make it consistent. We originally chose the word "plugins" because we didn't want people confusing this with things like node module dependencies.

Reply:
:+1: